tests/Doctrine/Tests/ORM/Functional/Ticket/GH9109Test.php

```diff
@@ -0,0 +1,201 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+use Doctrine\Common\Collections\Criteria;
+use Doctrine\ORM\Mapping\Column;
+use Doctrine\ORM\Mapping\Entity;
+use Doctrine\ORM\Mapping\GeneratedValue;
+use Doctrine\ORM\Mapping\Id;
+use Doctrine\ORM\Mapping\ManyToMany;
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+/**
+ * @group GH-9109
+ */
+class GH9109Test extends OrmFunctionalTestCase
+{
+    public function testIssue(): void
+    {
+        $userFirstName = 'GH9109Test';
+        $userLastName  = 'UserGH9109';
+        $productTitle  = 'Test product';
+
+        $this->_schemaTool->createSchema(
+            [
+                $this->_em->getClassMetadata(GH9109User::class),
+                $this->_em->getClassMetadata(GH9109Product::class),
+            ]
+        );
+
+        $userRepository = $this->_em->getRepository(GH9109User::class);
+
+        $user = new GH9109User();
+        $user->setFirstName($userFirstName);
+        $user->setLastName($userLastName);
+
+        $product = new GH9109Product();
+        $product->setTitle($productTitle);
+
+        $this->_em->persist($user);
+        $this->_em->persist($product);
+        $this->_em->flush();
+
+        $product->addBuyer($user);
+
+        $this->_em->persist($product);
+        $this->_em->flush();
+
+        $this->_em->clear();
+
+        $persistedProduct = $this->_em->find(GH9109Product::class, $product->getId());
+
+        // assert Product was persisted
+        self::assertInstanceOf(GH9109Product::class, $persistedProduct);
+        self::assertEquals($productTitle, $persistedProduct->getTitle());
+
+        // assert Product has a Buyer
+        $criteria = Criteria::create();
+        $count = $persistedProduct->getBuyers()->matching($criteria)->count();
+        self::assertEquals(1, $count);
+
+        // assert NOT QUOTED WORK with findOneBy
+        $user = $userRepository->findOneBy(['lastName' => $userLastName]);
+        self::assertInstanceOf(GH9109User::class, $user);
+        self::assertEquals($userLastName, $user->getLastName());
+
+        // assert NOT QUOTED WORK with Criteria
+        $criteria = Criteria::create();
+        $criteria->where($criteria->expr()->eq('lastName', $userLastName));
+        $matchedUserLastName = $persistedProduct->getBuyers()->matching($criteria)->first();
```

Comment:
This variable name is confusing: one might think it will be a name, when it really is a user. IMO you could just reuse `$user` here, like you did below

Reply:
fixed. Thanks for the sugestion